stride/utils/operators.py

```diff
@@ -29,27 +29,117 @@ def adjoint(self, d_mul, a, b, **kwargs):
         return d_a, d_b
 
 
+@mosaic.tessera
 class Concatenate(Operator):
-
-    def forward(self, *args, **kwargs):
-        axis = kwargs.pop('axis', 0)
+    """
+    Concatenate multiple StructuredData objects.
+
+    Parameters
+    ----------
+    start_end: 2d-array, optional
+        Array containing start and end indices pointing to each objects' position in the concatenated object.
+        Shape should be (num_objects, 2)
+    new_axis : bool, optional
+        Whether to concatenate on a new axis. Defaults to False.
+    axis : int, optional
+        Axis to perform the concatenation. Defaults to 0.
+
+    """
+    def __init__(self, *args, **kwargs):
+        self._start_end = kwargs.pop('start_end', None)
+        self.new_axis = kwargs.pop('new_axis', False)
+        self.axis = kwargs.pop('axis', 0)
+        super().__init__(*args, **kwargs)
+
+    async def forward(self, *args, **kwargs):
+        """
+        Parameters
+        ----------
+        args : Sequence[StructuredData]
+            Sequence of StructuredData objects to be combined.
+        start_end : 2d-array, optional
+            Array containing start and end indices pointing to each objects' position in the concatenated object.
+            Shape should be (num_objects, 2)
+        axis : int, optional
+            The axis to concatenate on. Defaults to 0.
+        new_axis : bool, optional
+            Whether to create a new axis when concatenating, or maintain the dimensions. Defaults to False.
+
+        Returns
+        -------
+        StructuredData
+            Concatenated data as a single StructuredData object.
+
+        """
+        new_axis = kwargs.pop('new_axis', None)
+        axis = kwargs.pop('axis', None)
+        start_end = kwargs.pop('start_end', None)
+
+        if new_axis is not None:
+            self.new_axis = new_axis
+
+        if axis is not None:
+            self.axis = axis
+
+        if start_end is not None:
+            # update if required
+            self._start_end = start_end
+        elif self._start_end is None:
+            # build start_end from the arguments
+            self.build_start_end(args)
 
         concat_data = [np.array(each.data) for each in args]
-        concat_data = np.stack(concat_data, axis=axis)
 
-        concat = StructuredData(name='concat', data=concat_data, grid=args[0].grid)
+        if self.new_axis:
+            concat_data = np.stack(concat_data, axis=self.axis)
+        else:
+            concat_data = np.vstack(concat_data)
 
-        return concat
+        concat = args[0].alike(name='concat_%s' % args[0].name, data=concat_data, shape=None, extended_shape=None, inner=None)
```

Comment:
is it worth checking that `args` has a `length > 0` before this line?

Reply:
Agreed

Reply:
Thought about this a little more - the `Concatenate` operator only really makes sense if `len(args) >= 2`, i.e. there is more than one `StructuredData` object to perform the concatenation. So should this check be used instead?

Reply:
Completed